### PR TITLE
[SPARK-30921][PySpark] Predicates on python udf should not be pushdown through Aggregate

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch proposed to skip predicates on PythonUDFs to be pushdown through Aggregate.

### Why are the changes needed?

The predicates on PythonUDFs cannot be pushdown through Aggregate. Pushed down predicates cannot be evaluate because PythonUDFs cannot be evaluated on Filter and cause error like:

```
Caused by: java.lang.UnsupportedOperationException: Cannot generate code for expression: mean(input[1, struct<bar:bigint>, true].bar)                                        
        at org.apache.spark.sql.catalyst.expressions.Unevaluable.doGenCode(Expression.scala:304)                                                                             
        at org.apache.spark.sql.catalyst.expressions.Unevaluable.doGenCode$(Expression.scala:303)                                                                            
        at org.apache.spark.sql.catalyst.expressions.PythonUDF.doGenCode(PythonUDF.scala:52)                                                                                 
        at org.apache.spark.sql.catalyst.expressions.Expression.$anonfun$genCode$3(Expression.scala:146)                                                                     
        at scala.Option.getOrElse(Option.scala:189)                                                                                                                          
        at org.apache.spark.sql.catalyst.expressions.Expression.genCode(Expression.scala:141)                                                                                
        at org.apache.spark.sql.catalyst.expressions.CastBase.doGenCode(Cast.scala:821)                                                                                      
        at org.apache.spark.sql.catalyst.expressions.Expression.$anonfun$genCode$3(Expression.scala:146)                                                                     
        at scala.Option.getOrElse(Option.scala:189)                                           
```

### Does this PR introduce any user-facing change?

Yes. Previously the predicates on PythonUDFs will be pushdown through Aggregate can cause error. After this change, the query can work.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @HyukjinKwon @BryanCutler

**Comment by HyukjinKwon** [comment]
Will take a look tomorrow

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -1190,7 +1190,8 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
     // Find all the aliased expressions in the aggregate list that don't include any actual
     // AggregateExpression, and create a map from the alias to the expression
     val aliasMap = plan.aggregateExpressions.collect {
-      case a: Alias if a.child.find(_.isInstanceOf[AggregateExpression]).isEmpty =>
+      case a: Alias if a.child.find(_.isInstanceOf[AggregateExpression]).isEmpty &&
+          a.child.find(_.isInstanceOf[PythonUDF]).isEmpty =>
```
Comment (BryanCutler) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1194]: could you check both instances in a single call to `find`? like `a.child.find(e => e.isInstanceOf[AggregateExpression] || e.isInstanceOf[PythonUDF]).isEmpty`

**Review by BryanCutler — COMMENTED** [review]
looks good to me, just a minor suggestion

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1194]: ok.

File: python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py [hunk 2]
```diff
@@ -491,6 +491,28 @@ def max_udf(v):
             agg2 = self.spark.sql("select max_udf(id) from table")
             assert_frame_equal(agg1.toPandas(), agg2.toPandas())
 
+    def test_no_predicate_pushdown_through(self):
+        from pyspark.sql.functions import monotonically_increasing_id, explode_outer
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py:495]: @viirya, I think we should better leave a comment like:

```
def test_case(self):
    # SPARK-12345: a short description of the test
    ...
```

per the new guide written recently (https://spark.apache.org/contributing.html)

File: python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py [hunk 3]
```diff
@@ -491,6 +491,28 @@ def max_udf(v):
...
+        @pandas_udf('float', PandasUDFType.GROUPED_AGG)
+        def mean(x):
+            return np.mean(x)
+
+        df = self.spark.createDataFrame([
+            Row(foo=[Row(bar=42), Row(bar=43), Row(bar=44)]),
+        ])
+
+        df_with_id = df.withColumn('id', monotonically_increasing_id())
+        exploded = df_with_id.select('id', explode_outer('foo').alias('foos'))
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py:507]: @viirya, is it necessary to use `monotonically_increasing_id` and `explode_outer`?  I know it's extracted from the reported case but I think we should just make it minimised for the fix.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -1188,9 +1188,10 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
 
   def getAliasMap(plan: Aggregate): AttributeMap[Expression] = {
     // Find all the aliased expressions in the aggregate list that don't include any actual
-    // AggregateExpression, and create a map from the alias to the expression
+    // AggregateExpression or PythonUDF, and create a map from the alias to the expression
     val aliasMap = plan.aggregateExpressions.collect {
-      case a: Alias if a.child.find(_.isInstanceOf[AggregateExpression]).isEmpty =>
+      case a: Alias if a.child.find(e => e.isInstanceOf[AggregateExpression] ||
+          e.isInstanceOf[PythonUDF]).isEmpty =>
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1194]: What about we use `PythonUDF.isGroupedAggPandasUDF(e)` instead? This is guaranteed in  https://github.com/apache/spark/blob/5866bc77d7703939e93c00f22ea32981d4ebdc6c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala#L213-L215

If we need another Python UDF that supports aggregation, we should touch there anyway.

**Comment by HyukjinKwon** [comment]
LGTM otherwise

File: python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py (hunk 3 above)
Comment (viirya) [reply python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py:507]: Ah, this comes from reported test code. Let me simplify it a bit. :)

File: python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py (hunk 2 above)
Comment (viirya) [reply python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py:495]: Sure. Thanks for reminding me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 4 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1194]: Ok. Sounds good.

File: python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py [hunk 5]
```diff
@@ -491,6 +491,23 @@ def max_udf(v):
             agg2 = self.spark.sql("select max_udf(id) from table")
             assert_frame_equal(agg1.toPandas(), agg2.toPandas())
 
+    def test_no_predicate_pushdown_through(self):
+        # SPARK-12345: We should not pushdown predicates of PythonUDFs through Aggregate.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py:495]: JIRA I'd:).

Comment (viirya) [reply python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py:495]: Oops! :)

**Comment by HyukjinKwon** [comment]
Merged to master, branch-3.0 and branch-2.4.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master, branch-3.0 and branch-2.4.

**Pull request closed without merging**